table/update_schema.go

```diff
@@ -0,0 +1,532 @@
+		return us
+	}
+
+	// check promotion
+	if !allowedPromotion(field.Type, newType) {
+		panic("Cannot update type of column: " + field.Name + ": " + newType.String())
+	}
+
+	us.updates[field.ID].Type = newType
+
+	return us
+}
+
+func (us *UpdateSchema) UpdateColumnDoc(name string, doc string) *UpdateSchema {
+	field := us.findForUpdate(name)
+	if field == nil {
+		panic("Cannot update type of missing column: " + name)
+	}
+
+	for _, id := range us.deletes {
+		if id == field.ID {
+			panic("Cannot update a column that will be deleted: " + field.Name)
+		}
+	}
+
+	if field.Doc == doc {
+		return us
+	}
+
+	us.updates[field.ID].Doc = doc
+
+	return us
+}
+
+func (us *UpdateSchema) UpdateColumnDefault(name string, defaultValue any) *UpdateSchema {
+	field := us.findForUpdate(name)
+	if field == nil {
+		panic("Cannot update default value of missing column: " + name)
+	}
+
+	for _, id := range us.deletes {
+		if id == field.ID {
+			panic("Cannot update a column that will be deleted: " + field.Name)
+		}
+	}
+
+	if field.InitialDefault == defaultValue {
+		return us
+	}
+
+	us.updates[field.ID].InitialDefault = defaultValue
+
+	return us
+}
+
+// RequireColumn changes an optional column to required.
+func (us *UpdateSchema) RequireColumn(name string) *UpdateSchema {
+	us.internalUpdateColumnRequirement(name, true)
+
+	return us
+}
+
+// MakeColumnOptional changes a required column to optional.
+func (us *UpdateSchema) MakeColumnOptional(name string) *UpdateSchema {
+	us.internalUpdateColumnRequirement(name, false)
+
+	return us
+}
+
+func (us *UpdateSchema) findForUpdate(name string) *iceberg.NestedField {
+	existing := us.findField(name)
+	if existing != nil {
+		if update, ok := us.updates[existing.ID]; ok {
+			return update
+		}
+
+		// adding to updates
+		us.updates[existing.ID] = existing
+
+		return existing
+	}
+
+	addedID, ok := us.addedNameToID[name]
+	if ok {
+		return us.updates[addedID]
+	}
+
+	return nil
+}
+
+func (us *UpdateSchema) Apply() *iceberg.Schema {
+	return us.applyChanges()
+}
+
+func (su *UpdateSchema) isAdded(name string) bool {
+	_, ok := su.addedNameToID[name]
+
+	return ok
+}
+
+func (us *UpdateSchema) assignNewColumnID() int {
+	next := us.lastColumnID + 1
+	us.lastColumnID = next
+
+	return next
+}
+
+func (us *UpdateSchema) findField(name string) *iceberg.NestedField {
+	if us.caseSensitive {
+		field, ok := us.schema.FindFieldByName(name)
+		if !ok {
+			return nil
+		}
+
+		return &field
+	}
+	field, ok := us.schema.FindFieldByNameCaseInsensitive(name)
+	if !ok {
+		return nil
+	}
+
+	return &field
+}
+
+func allowedPromotion(oldType, newType iceberg.Type) bool {
+	switch old := oldType.(type) {
+	case iceberg.PrimitiveType:
+		switch old.Type() {
+		case "int":
+			return newType.Type() == "long"
+		case "float":
+			return newType.Type() == "double"
+		case "string":
+			return newType.Type() == "binary" // widening
+		case "fixed":
+			return newType.Type() == "binary"
+		case "time":
```

Comment:
Is time -> timestamp promotion allowed? From spec, it appears to me that only date -> timestamp promotion is allowed.